subsys/net/lib/mqtt_helper/mqtt_helper.c

```diff
@@ -383,31 +371,48 @@ static int broker_init(struct sockaddr_storage *broker,
 	addr = result;
 
 	while (addr != NULL) {
-		if (addr->ai_addrlen == sizeof(struct sockaddr_in)) {
+		if (addr->ai_family == AF_INET6) {
+			struct sockaddr_in6 *broker6 = ((struct sockaddr_in6 *)broker);
+
+			broker6->sin6_family = addr->ai_family;
+			broker6->sin6_port = htons(CONFIG_MQTT_HELPER_PORT);
+			net_ipaddr_copy(&broker6->sin6_addr,
+					&((struct sockaddr_in6 *)addr->ai_addr)->sin6_addr);
+
+			inet_ntop(addr->ai_family, &broker6->sin6_addr, addr_str, sizeof(addr_str));
+			LOG_INF("IPv6 Address found %s (%s)", addr_str,
+				net_family2str(addr->ai_family));
+
+			found = true;
+			break;
+		} else if (addr->ai_family == AF_INET) {
 			struct sockaddr_in *broker4 = ((struct sockaddr_in *)broker);
-			char ipv4_addr[INET_ADDRSTRLEN];
 
-			broker4->sin_addr.s_addr =
-				((struct sockaddr_in *)addr->ai_addr)->sin_addr.s_addr;
-			broker4->sin_family = AF_INET;
+			broker4->sin_family = addr->ai_family;
 			broker4->sin_port = htons(CONFIG_MQTT_HELPER_PORT);
+			net_ipaddr_copy(&broker4->sin_addr,
+					&((struct sockaddr_in *)addr->ai_addr)->sin_addr);
 
-			inet_ntop(AF_INET, &broker4->sin_addr.s_addr, ipv4_addr,
-				  sizeof(ipv4_addr));
-			LOG_DBG("IPv4 Address found %s", ipv4_addr);
+			inet_ntop(addr->ai_family, &broker4->sin_addr, addr_str, sizeof(addr_str));
+			LOG_INF("IPv4 Address found %s (%s)", addr_str,
+				net_family2str(addr->ai_family));
+
+			found = true;
 			break;
+		} else {
+			LOG_DBG("Unknown address family %d", (unsigned int)addr->ai_family);
 		}
 
-		LOG_DBG("ai_addrlen is %u, while it should be %u",
-			(unsigned int)addr->ai_addrlen,
-			(unsigned int)sizeof(struct sockaddr_in));
-
 		addr = addr->ai_next;
-		break;
 	}
 
 	freeaddrinfo(result);
 
+	if (!found) {
+		LOG_ERR("No address found for %s", conn_params->hostname.ptr);
+		return -1;
```

Comment:
Typically, some errno code that represents the error the best the error is used.
`-EFAULT`
`-ECANCELED`
`-ECHILD` (used in `nrf_cloud_transport.c`)

Will this check ever get triggered? Doesn't `getaddrinfo` return an error if no address is found?

Anyway, depending on the return value you choose it would be nice to include documentation for it in `mqtt_helper.h` for the `mqtt_helper_connect` function.

Reply:
Thanks. I checked and getaddrinfo would return an error if no address found, so I will remove this check.